src/elementpenalty.jl

```diff
@@ -4,69 +4,66 @@ Penalties that are applied element-wise.
 abstract type ElementPenalty <: Penalty end
 abstract type ConvexElementPenalty <: ElementPenalty end  # only these have prox method
 
-# Make broadcast work for ElementPenalty
-Base.getindex(p::ElementPenalty, idx) = p
-Base.size(::ElementPenalty) = ()
-
 #-------------------------------------------------------------------------------# methods
-value{T}(p::ElementPenalty, θ::T, s::T)     = s * value(p, θ)
-value{T}(p::ElementPenalty, θ::AA{T})       = sum(x -> value(p, x), θ)
-value{T}(p::ElementPenalty, θ::AA{T}, s::T) = sum(x -> value(p, x, s), θ)
-function value{T}(p::ElementPenalty, θ::AA{T}, s::AA{T})
+value(p::ElementPenalty, θ::Number, s::Number)       = s * value(p, θ)
+value(p::ElementPenalty, θ::AA{<:Number})            = sum(x -> value(p, x), θ)
+value(p::ElementPenalty, θ::AA{<:Number}, s::Number) = sum(x -> value(p, x, s), θ)
+function value{T <: Number,S <: Number}(p::ElementPenalty, θ::AA{T}, s::AA{S})
     @assert size(θ) == size(s)
-    result = zero(T)
-    for i in eachindex(θ)
-        @inbounds result += value(p, θ[i], s[i])
+    # TODO: make this work: Core.Inference.return_type(value, (typeof(p), T, S))
```

Comment:
well, right now it uses `result = zero(value(p, first(θ), first(s)))` which calls `value`, even though we just want the return type. I though `Core.inference.return_type` could avoid that. but it doesn't look like it

ref: https://discourse.julialang.org/t/using-core-inference-return-type/2945